ipam/allocator.go

```diff
@@ -0,0 +1,610 @@
+		if ones < minNetSizeV6 {
+			return nil, ErrInvalidSubnet
+		}
+		if ones < minNetSizeV6Eff {
+			newMask := net.CIDRMask(minNetSizeV6Eff, bits)
+			return &net.IPNet{IP: subnet.IP, Mask: newMask}, nil
+		}
+	} else {
+		if ones < minNetSize {
+			return nil, ErrInvalidSubnet
+		}
+	}
+	return subnet, nil
+}
+
+// Checks whether the passed subnet is a superset or subset of any of the subset in the db
+func (a *Allocator) contains(space AddressSpace, subInfo *SubnetInfo) bool {
+	a.Lock()
+	defer a.Unlock()
+	for k, v := range a.subnets {
+		if space == k.addressSpace {
+			if subInfo.Subnet.Contains(v.Subnet.IP) ||
+				v.Subnet.Contains(subInfo.Subnet.IP) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Splits the passed subnet into N internal subnets with host size equal to internalHostSize.
+// If the subnet's host size is equal to or smaller than internalHostSize, there won't be any
+// split and the return list will contain only the passed subnet.
+func getInternalSubnets(inSubnet *net.IPNet, internalHostSize int) ([]*net.IPNet, error) {
+	var subnetList []*net.IPNet
+
+	// Sanity check and size adjustment for v6
+	subnet, err := adjustAndCheckSubnetSize(inSubnet)
+	if err != nil {
+		return subnetList, err
+	}
+
+	// Get network/host subnet information
+	netBits, bits := subnet.Mask.Size()
+	hostBits := bits - netBits
+
+	extraBits := hostBits - internalHostSize
+	if extraBits <= 0 {
+		subnetList = make([]*net.IPNet, 1)
+		subnetList[0] = subnet
+	} else {
+		// Split in smaller internal subnets
+		numIntSubs := 1 << uint(extraBits)
+		subnetList = make([]*net.IPNet, numIntSubs)
+
+		// Construct one copy of the internal subnets's mask
+		intNetBits := bits - internalHostSize
+		intMask := net.CIDRMask(intNetBits, bits)
+
+		// Construct the prefix portion for each internal subnet
+		for i := 0; i < numIntSubs; i++ {
+			intIP := make([]byte, len(subnet.IP))
+			copy(intIP, subnet.IP) // IPv6 is too big, just work on the extra portion
+			addIntToIP(intIP, i<<uint(internalHostSize))
+			subnetList[i] = &net.IPNet{IP: intIP, Mask: intMask}
+		}
+	}
+	return subnetList, nil
+}
+
+// RemoveSubnet removes the subnet from the specified address space
+func (a *Allocator) RemoveSubnet(addrSpace AddressSpace, subnet *net.IPNet) error {
+	if addrSpace == "" {
+		return ErrInvalidAddressSpace
+	}
+	if subnet == nil {
+		return ErrInvalidSubnet
+	}
+retry:
+	// Look for the respective subnet configuration data
+	// Remove it along with the internal subnets
+	subKey := subnetKey{addrSpace, subnet.String(), ""}
+	a.Lock()
+	_, ok := a.subnets[subKey]
+	a.Unlock()
+	if !ok {
+		return ErrSubnetNotFound
+	}
+
+	// Remove config
+	a.Lock()
+	delete(a.subnets, subKey)
+	a.Unlock()
+
+	// Sync to datastore
+	err := a.writeToStore()
+	if err != nil {
+		// Update to latest in any error case
+		a.readFromStore()
+		if _, ok := err.(types.RetryError); !ok {
+			return types.InternalErrorf("subnet removal failed because of %s", err.Error())
```

Comment:
Same comment as above for `AddSubnet` where the read fails you need to rollback to the previous state